File: RotateList.cc
#include <iostream>
using namespace std;

struct ListNode {
      int val;
      ListNode *next;
      ListNode(int x) : val(x), next(NULL) {}
 }; 

ListNode *rotateRight(ListNode *head, int k) {
    int len = 0;
    ListNode* tail = head;
    if (!head)
        return head;
    for (; tail->next; tail = tail->next, ++len );
    ++len;
    k = k % len;
    if (k == 0)
        return head;

    int count = len - k;
    ListNode* prev = head;
    for (int i = 0; i < count-1; ++i)
        prev = prev->next;
    tail->next = head;
    head = prev->next;
    prev->next = NULL;

    return head;
}

int main(int argc, char** argv)
{
    return 0;
}

File: RotateList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class RotateList {
    public ListNode rotateRight(ListNode head, int k) {
        if (k <= 0 || head == null) {
            return head;
        }
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode pre = dummy;
        ListNode cur = dummy;
        int len = 0;
        for(; cur.next != null; cur = cur.next, ++len);
        k %= len;
        if (k == 0) {
            return head;
        }
        cur.next = head;
        for(int i = 0; i < len - k; pre = pre.next, ++i);
        head = pre.next;
        pre.next = null;
        return head;
    }
}